### ISSUE TITLE
ZeroShot

### ISSUE THREAD

**Issue by sucduit** [MaartenGr/BERTopic#1775]
I use ZeroShot to classify my documents. 
topic_model_zero1 = BERTopic(
    hdbscan_model=hdbscan_model,
    embedding_model="thenlper/gte-small", 
    min_topic_size=15,
    zeroshot_topic_list=zeroshot_topic_list,
    zeroshot_min_similarity=.75,
    representation_model=KeyBERTInspired()
)
topics, probs= topic_model_1.fit_transform(texts_zeroshot)
topics, probs = topic_model_1.transform(texts_zeroshot)
Zero_1_Doc_Info=topic_model_1.get_document_info(texts_zeroshot)
topic_distr, _ = topic_model_1.approximate_distribution(texts_zeroshot)

I can get the following information:
Document | Topic | Name | Representation | Representative_Docs | Top_n_words| Representative_document
The representative document is shown as false and there is no probability distribution for each document.
My boss asked for the topic probability distribution of each document and I cannot figure out how to do it. Please check my code and help me to find the information I need. Thanks!

**Comment by MaartenGr** [comment]
To get the probabilities, you should set `calculate_probalities=True` when initializing your BERTopic model. You could also set `topic_model_1.calculate_probabilities = True` after training. Note that the representative documents are not saved with zeroshot topic modeling and you would have to calculate those yourself. You can check existing issues for the code.

**Comment by sucduit** [comment]
I added calculate_probalities=True and also did topic_model_1.calculate_probabilities = True after fitting the model. But I still cannot get the probabilities. What I should do to calculate the probability. Please help. Thanks

**Comment by MaartenGr** [comment]
Could you share your **full code**? It's not clear how some of the variables are created.

**Comment by sucduit** [comment]
Sure. The following is my full code:

from bertopic import BERTopic
from bertopic.representation import KeyBERTInspired
from transformers.pipelines import pipeline
from hdbscan import HDBSCAN
import pandas as pd
data = pd.read_excel('C:/Users/messages.xlsx')
texts_zeroshot = data['Doc']
zeroshot_topic_list = ["corporate culture", "investment", ...."diversity"]
#I listed around 20 possible related topics
hdbscan_model = HDBSCAN(min_cluster_size=15, metric='euclidean',
cluster_selection_method='eom', prediction_data=True)
topic_model= BERTopic(
    embedding_model="thenlper/gte-small",
    hdbscan_model=hdbscan_model,
    min_topic_size=15,
    zeroshot_topic_list=zeroshot_topic_list,
    zeroshot_min_similarity=.85,
    representation_model=KeyBERTInspired(),
    calculate_probabilities=True,
    verbose=True
)
topics= topic_model.fit(texts_zeroshot)
topic_model.calculate_probabilities = True
topics, probs = topic_model.transform(texts_zeroshot)
Doc_Info=topic_model.get_document_info(texts_zeroshot)
Doc_Info.head(5)
Topic_Info=topic_model.get_topic_info()
Topic_Info

From the above code, I got the following information :
Document Topic Name Representation Representative_Docs Top_n_words
Representative_document

I got 26 classified topics. Among the information returned,
Representative_document, all return "False" and there is no probability
value.  I need mining topics in part of my dissertation. Our previous Phd
students all use traditional methods like LDA in theirs. I am the first one
using Bertopic and my chair is quite open to it but he asked for some
information based on his previous supervised paper experience. Probability
is one of them and If you can help me with it, I will appreciate it very
much.

Best,
Ellen

**Comment by MaartenGr** [comment]
Hmmm, I can't seem to reproduce the issue, when I run the following the `probs` contains the probabilities for all documents:

```python
from bertopic import BERTopic
from bertopic.representation import KeyBERTInspired
from transformers.pipelines import pipeline
from hdbscan import HDBSCAN
import pandas as pd
from sentence_transformers import SentenceTransformer

from datasets import load_dataset

# Prepare data
dataset = load_dataset("CShorten/ML-ArXiv-Papers")["train"]
abstracts = dataset["abstract"][:5_000]

# Precaculate embeddings
embedding_model = SentenceTransformer("thenlper/gte-small")
embeddings = embedding_model.encode(abstracts)

# Define zeroshot topics
zeroshot_topic_list = ["Clustering", "Topic Modeling", "Large Language Models"]

# Prepare HDBSCAN model
hdbscan_model = HDBSCAN(min_cluster_size=15, metric='euclidean', cluster_selection_method='eom', prediction_data=True)

# Fit model
topic_model= BERTopic(
    embedding_model=embedding_model,
    hdbscan_model=hdbscan_model,
    min_topic_size=15,
    zeroshot_topic_list=zeroshot_topic_list,
    zeroshot_min_similarity=.85,
    representation_model=KeyBERTInspired(),
    calculate_probabilities=True,
    verbose=True
)
topic_model = topic_model.fit(abstracts, embeddings)
topics, probs = topic_model.transform(abstracts, embeddings)
```

In other words, the `probs` contains the probability distribution that you want for each document. Their order relates to the input documents. In my example, `probs[0]` relates to `abstracts[0]`,  `probs[1]` relates to `abstracts[1]`, `probs[2]` relates to `abstracts[2]`, etc.

**Comment by sucduit** [comment]
Thank you for the reply. I use your code to do it and then to get the
information as followings:
Doc_Info=topic_model.get_document_info(texts_zeroshot)
Doc_Info.head(5)
There is still no probability information as the screenshot below. If the
probs contain the probability distribution here but didn't show in the
results, is there any way in which I can get the probs information using
some code? If you can offer some help, I appreciate it very much.
[image: image.png]

**Comment by sucduit** [comment]
I use  probs_df=pd.DataFrame(probs) and did get a probability distribution
for each document but the weird thing is the probabilities for each topic
are very large. For example, the document 0's topic 0 has prob 0,931867 and
the document 0's topic 1 is 0.953366...... Are those probabilities supposed
to be added up to 1? Why are they so large? Is this the reason doc
information won't give me prob information? If so, How can I fix it? Thanks.
[image: image.png]

**Comment by MaartenGr** [comment]
Most likely you would need to run `probs.T` to get the transposed probability matrix. I think you are mistaking rows for columns.
